File: flutter-example/dsrv-wallet-sdk-flutter/ios/Classes/DsrvWalletSdkPlugin.swift
import Flutter
import UIKit
import Security
import dsrv_wallet_sdk_ios

/// Flutter ↔ native DSRVWallet 브릿지.
/// - 정방향: Dart invokeMethod → DSRVWallet 호출
/// - 역방향: AuthHandler.requestChallenge → Dart `onRequestChallenge`
public class DsrvWalletSdkPlugin: NSObject, FlutterPlugin {

    private let channel: FlutterMethodChannel

    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.dsrv.wallet.sdk/api",
            binaryMessenger: registrar.messenger()
        )
        let instance = DsrvWalletSdkPlugin(channel: channel)
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    /// FlutterError 는 Swift `Error` 를 conform 하지 않아 직접 throw 불가 — 내부 errno 로 변환 후
    /// `handle` 의 catch 에서 FlutterError 로 매핑한다.
    private struct PluginError: Error {
        let code: String
        let message: String
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = (call.arguments as? [String: Any]) ?? [:]
        Task {
            do { try await dispatch(call.method, args, result) }
            catch let e as PluginError {
                result(FlutterError(code: e.code, message: e.message, details: nil))
            }
            catch {
                result(FlutterError(code: "INTERNAL",
                                    message: error.localizedDescription, details: nil))
            }
        }
    }

    /// Required string arg helper — Android `MethodCall.arg<T>()` 와 동치. missing 또는 빈 문자열을
    /// silent default ("") 로 coalesce 하지 않고 즉시 INVALID_ARGUMENT 로 surface 한다.
    private func requiredString(_ args: [String: Any], _ key: String) throws -> String {
        guard let v = args[key] as? String, !v.isEmpty else {
            throw PluginError(code: "INVALID_ARGUMENT",
                              message: "missing or empty arg: \(key)")
        }
        return v
    }

    private func dispatch(_ method: String, _ args: [String: Any], _ result: @escaping FlutterResult) async throws {
        switch method {
        case "initialize":
            let credential = UserCredential(
                type: credentialType(args["credentialType"] as? String),
                value: try requiredString(args, "credentialValue"),
                provider: args["provider"] as? String ?? ""
            )
            let r = await DSRVWallet.initialize(
                sdkId: try requiredString(args, "sdkId"),
                userCredential: credential,
                authHandler: DartAuthHandler(channel: channel),
                baseUrl: try requiredString(args, "baseUrl")
            )
            reply(r, result) { _ in nil }

        case "isInitialized":
            result(DSRVWallet.isInitialized)

        case "reset":
            await DSRVWallet.reset()
            result(nil)

        case "createAccount":
            let r = await DSRVWallet.createAccount(label: try requiredString(args, "label"))
            reply(r, result) { ["accountId": $0.accountId, "label": $0.label] }

        case "getAccountList":
            let r = await DSRVWallet.getAccountList()
            reply(r, result) { $0.map(self.accountMap) }

        case "getChainList":
            let r = await DSRVWallet.getChainList()
            reply(r, result) { $0.map(self.chainMap) }

        case "createAddress":
            let r = await DSRVWallet.createAddress(
                accountId: try requiredString(args, "accountId"),
                chainType: try requiredString(args, "chainType"),
                label: args["label"] as? String
            )
            reply(r, result) { ["publicKey": $0.publicKey, "address": $0.address] }

        case "transfer":
            let r = await DSRVWallet.transfer(
                address: try requiredString(args, "address"),
                chainId: try requiredString(args, "chainId"),
                asset: parseAsset(args["asset"] as? [String: Any] ?? [:]),
                recipient: try requiredString(args, "recipient"),
                amount: try requiredString(args, "amount")
            )
            reply(r, result) { ["txHash": $0.txHash] }

        case "buildTx":
            let r = await DSRVWallet.buildTx(
                address: try requiredString(args, "address"),
                chainId: try requiredString(args, "chainId"),
                asset: parseAsset(args["asset"] as? [String: Any] ?? [:]),
                recipient: try requiredString(args, "recipient"),
                amount: try requiredString(args, "amount")
            )
            reply(r, result) {
                ["txId": $0.txId, "signId": $0.signId, "messageHash": $0.messageHash, "type": $0.type]
            }

        case "broadcastTx":
            let r = await DSRVWallet.broadcastTx(
                address: try requiredString(args, "address"),
                txId: try requiredString(args, "txId")
            )
            reply(r, result) { ["txHash": $0.txHash] }

        case "sign":
            let r = await DSRVWallet.sign(
                address: try requiredString(args, "address"),
                hashedMessage: try requiredString(args, "hashedMessage"),
                signId: try requiredString(args, "signId"),
                messageType: try requiredString(args, "messageType")
            )
            reply(r, result) { ["r": $0.r, "s": $0.s, "v": $0.v] }

        case "delegate":
            let r = await DSRVWallet.delegate(address: try requiredString(args, "address"))
            reply(r, result) { $0.map(self.chainTxResultMap) }

        case "revoke":
            let r = await DSRVWallet.revoke(address: try requiredString(args, "address"))
            reply(r, result) { $0.map(self.chainTxResultMap) }

        case "approve":
            let r = await DSRVWallet.approve(address: try requiredString(args, "address"))
            reply(r, result) { $0.map(self.chainTxResultMap) }

        case "backup":
            let r = await DSRVWallet.backup()
            reply(r, result) { _ in nil }

        case "restore":
            let r = await DSRVWallet.restore()
            reply(r, result) { $0.map(self.restoredMap) }

        case "clearDeviceKeyForDebug":
            // 저장된 App Attest keyId 삭제 → 다음 init 시 새 키 생성·재attest.
            // (keychain 은 앱 삭제로도 안 지워지므로 디버그용으로 직접 제거)
            SecItemDelete([
                kSecClass as String: kSecClassGenericPassword,
                kSecAttrService as String: "com.dsrv.wallet.sdk.attest-key-id",
            ] as CFDictionary)
            result(nil)

        case "dumpBackupForDebug":
            result(DSRVWallet.dumpBackupForDebug())

        case "clearBackupForDebug":
            DSRVWallet.clearBackupForDebug()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - WalletResult → FlutterResult

    private func reply<T>(_ r: WalletResult<T>, _ result: @escaping FlutterResult, _ transform: (T) -> Any?) {
        switch r {
        case .success(let data): result(transform(data))
        case .failure(let error):
            result(FlutterError(code: "\(error.code)", message: error.description, details: nil))
        }
    }

    // MARK: - 직렬화 helper

    private func parseAsset(_ map: [String: Any]) -> TransferAsset {
        if (map["type"] as? String) == "erc20", let token = map["tokenAddress"] as? String {
            return .erc20(tokenAddress: token)
        }
        return .native
    }

    private func credentialType(_ wire: String?) -> CredentialType {
        switch wire {
        case "OAUTH_TOKEN": return .oauthToken
        case "IDP_TOKEN": return .idpToken
        default: return .userId
        }
    }

    private func accountMap(_ a: AccountInfo) -> [String: Any?] {
        return [
            "accountId": a.accountId,
            "label": a.label,
            "addresses": a.addresses.map { [
                "accountId": $0.accountId,
                "addressId": $0.addressId,
                "address": $0.address,
                "publicKey": $0.publicKey,
                "label": $0.label,
                "chainType": $0.chainType,
            ] },
        ]
    }

    private func chainMap(_ c: ChainInfo) -> [String: Any] {
        return ["chainId": c.chainId, "name": c.name, "chainType": c.chainType, "networkType": c.networkType]
    }

    private func restoredMap(_ r: RestoredKey) -> [String: Any?] {
        return ["address": r.address, "success": r.success, "error": r.error]
    }

    private func chainTxResultMap(_ r: ChainTxResult) -> [String: Any?] {
        return [
            "chainId": r.chainId,
            "outcome": r.outcome,
            "txHash": r.txHash,
            "errorMessage": r.errorMessage,
        ]
    }
}

/// AuthHandler.requestChallenge 를 Dart 로 역방향 위임.
final class DartAuthHandler: AuthHandler {
    private let channel: FlutterMethodChannel
    init(channel: FlutterMethodChannel) { self.channel = channel }

    func requestChallenge(request: ChallengeRequest) async -> ChallengeResult {
        let reqMap: [String: Any] = [
            "sdkId": request.sdkId,
            "appId": request.appId,
            "credentialType": request.userCredential.type.rawValue,
            "credentialValue": request.userCredential.value,
            "provider": request.userCredential.provider,
            "deviceInfo": [
                "keyId": request.deviceInfo.keyId as Any,
                "publicKey": request.deviceInfo.publicKey as Any,
                "model": request.deviceInfo.model,
                "osVersion": request.deviceInfo.osVersion,
                "isVirtual": request.deviceInfo.isVirtual,
                "attestationObject": request.deviceInfo.attestationObject as Any,
                "platform": "IOS",
            ],
        ]

        let reply: Any? = await withCheckedContinuation { cont in
            DispatchQueue.main.async {
                self.channel.invokeMethod("onRequestChallenge", arguments: reqMap) { res in
                    cont.resume(returning: res)
                }
            }
        }

        guard let map = reply as? [String: Any], (map["success"] as? Bool) == true,
              let challenge = map["challenge"] as? String else {
            let err = (reply as? [String: Any])?["error"] as? String ?? "Challenge failed"
            return .failure(error: err)
        }
        return .success(challenge: challenge)
    }
}

File: flutter-example/dsrv-wallet-sdk-flutter/android/src/main/kotlin/com/dsrv/wallet/flutter/DsrvWalletSdkPlugin.kt
package com.dsrv.wallet.flutter

import android.content.Context
import android.os.Handler
import android.os.Looper
import androidx.fragment.app.FragmentActivity
import com.dsrv.wallet.sdk.AuthHandler
import com.dsrv.wallet.sdk.ChallengeRequest
import com.dsrv.wallet.sdk.ChallengeResult
import com.dsrv.wallet.sdk.CredentialType
import com.dsrv.wallet.sdk.DSRVWallet
import com.dsrv.wallet.sdk.TransferAsset
import com.dsrv.wallet.sdk.UserCredential
import com.dsrv.wallet.sdk.WalletResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlin.coroutines.resume

/**
 * Flutter ↔ native DSRVWallet 브릿지.
 * - 정방향: Dart invokeMethod → DSRVWallet 호출
 * - 역방향: AuthHandler.requestChallenge → Dart `onRequestChallenge`
 */
class DsrvWalletSdkPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private lateinit var appContext: Context
    private var activity: FragmentActivity? = null
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        appContext = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "com.dsrv.wallet.sdk/api")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity as? FragmentActivity
    }

    override fun onDetachedFromActivity() { activity = null }
    override fun onReattachedToActivityForConfigChanges(b: ActivityPluginBinding) {
        activity = b.activity as? FragmentActivity
    }
    override fun onDetachedFromActivityForConfigChanges() { activity = null }

    override fun onMethodCall(call: MethodCall, result: Result) {
        scope.launch {
            try {
                handle(call, result)
            } catch (e: Exception) {
                result.error("9001", e.message ?: "Unknown", null)
            }
        }
    }

    private suspend fun handle(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> {
                val sdkId = call.arg<String>("sdkId")
                val credential = UserCredential(
                    type = CredentialType.valueOf(
                        (call.argument<String>("credentialType") ?: "USER_ID")
                            .let { wireToEnum(it) }
                    ),
                    value = call.arg("credentialValue"),
                    provider = call.argument<String>("provider") ?: ""
                )
                val r = DSRVWallet.initialize(
                    appContext, sdkId, credential, dartAuthHandler(),
                    baseUrl = call.arg("baseUrl"),
                    cloudProjectNumber = call.argument<Number>("cloudProjectNumber")?.toLong()
                )
                r.reply(result) { null }
            }

            "isInitialized" -> result.success(DSRVWallet.isInitialized)

            "reset" -> {
                DSRVWallet.reset()
                result.success(null)
            }

            "createAccount" -> DSRVWallet.createAccount(call.arg("label"))
                .reply(result) { mapOf("accountId" to it.accountId, "label" to it.label) }

            "getAccountList" -> DSRVWallet.getAccountList()
                .reply(result) { list -> list.map { accountMap(it) } }

            "getChainList" -> DSRVWallet.getChainList()
                .reply(result) { list -> list.map { chainMap(it) } }

            "createAddress" -> DSRVWallet.createAddress(
                accountId = call.arg("accountId"),
                chainType = call.arg("chainType"),
                label = call.argument<String>("label")
            ).reply(result) { mapOf("publicKey" to it.publicKey, "address" to it.address) }

            "transfer" -> DSRVWallet.transfer(
                address = call.arg("address"),
                chainId = call.arg("chainId"),
                asset = parseAsset(call.arg("asset")),
                recipient = call.arg("recipient"),
                amount = call.arg("amount")
            ).reply(result) { mapOf("txHash" to it.txHash) }

            "buildTx" -> DSRVWallet.buildTx(
                address = call.arg("address"),
                chainId = call.arg("chainId"),
                asset = parseAsset(call.arg("asset")),
                recipient = call.arg("recipient"),
                amount = call.arg("amount")
            ).reply(result) {
                mapOf(
                    "txId" to it.txId,
                    "signId" to it.signId,
                    "messageHash" to it.messageHash,
                    "type" to it.type
                )
            }

            "broadcastTx" -> DSRVWallet.broadcastTx(
                address = call.arg("address"),
                txId = call.arg("txId")
            ).reply(result) { mapOf("txHash" to it.txHash) }

            "sign" -> DSRVWallet.sign(
                address = call.arg("address"),
                hashedMessage = call.arg("hashedMessage"),
                signId = call.arg("signId"),
                messageType = call.arg("messageType")
            ).reply(result) { mapOf("r" to it.r, "s" to it.s, "v" to it.v) }

            "delegate" -> DSRVWallet.delegate(call.arg("address"))
                .reply(result) { list -> list.map(::chainTxResultMap) }

            "revoke" -> DSRVWallet.revoke(call.arg("address"))
                .reply(result) { list -> list.map(::chainTxResultMap) }

            "approve" -> DSRVWallet.approve(call.arg("address"), call.arg("amount"))
                .reply(result) { list -> list.map(::chainTxResultMap) }

            "backup" -> {
                val act = activity ?: return result.error(
                    "4201", "FragmentActivity 없음 (FlutterFragmentActivity 필요)", null)
                DSRVWallet.backup(act).reply(result) { null }
            }

            "restore" -> {
                val act = activity ?: return result.error(
                    "4202", "FragmentActivity 없음 (FlutterFragmentActivity 필요)", null)
                DSRVWallet.restore(act)
                    .reply(result) { list -> list.map { restoredMap(it) } }
            }

            // Android 는 Play Integrity(stateless)라 저장된 attest 키가 없음 → no-op
            "clearDeviceKeyForDebug" -> result.success(null)

            "dumpBackupForDebug" -> result.success(DSRVWallet.dumpBlockStoreForDebug())
            "clearBackupForDebug" -> { DSRVWallet.clearBackupForDebug(); result.success(null) }

            else -> result.notImplemented()
        }
    }

    // MARK: - AuthHandler 역방향 콜백

    private fun dartAuthHandler() = object : AuthHandler {
        override suspend fun requestChallenge(request: ChallengeRequest): ChallengeResult {
            val reply = invokeDart("onRequestChallenge", challengeReqMap(request))
            val map = reply as? Map<*, *>
                ?: return ChallengeResult.Failure("Invalid challenge reply")
            val success = map["success"] as? Boolean ?: false
            return if (success) {
                ChallengeResult.Success(map["challenge"] as? String ?: "")
            } else {
                ChallengeResult.Failure(map["error"] as? String ?: "Challenge failed")
            }
        }
    }

    /** native → Dart 호출을 메인 스레드에서 수행하고 suspend 로 결과 대기. */
    private suspend fun invokeDart(method: String, args: Any?): Any? =
        suspendCancellableCoroutine { cont ->
            mainHandler.post {
                channel.invokeMethod(method, args, object : Result {
                    override fun success(r: Any?) { if (cont.isActive) cont.resume(r) }
                    override fun error(c: String, m: String?, d: Any?) {
                        if (cont.isActive) cont.resume(mapOf("success" to false, "error" to m))
                    }
                    override fun notImplemented() {
                        if (cont.isActive) cont.resume(mapOf("success" to false, "error" to "notImplemented"))
                    }
                })
            }
        }

    // MARK: - 직렬화 helper

    private fun <T> WalletResult<T>.reply(result: Result, transform: (T) -> Any?) {
        when (this) {
            is WalletResult.Success -> result.success(transform(data))
            is WalletResult.Failure -> result.error(error.code.toString(), error.message, null)
        }
    }

    private fun parseAsset(map: Map<String, Any?>): TransferAsset =
        when (map["type"]) {
            "erc20" -> TransferAsset.Erc20(map["tokenAddress"] as String)
            else -> TransferAsset.Native
        }

    private fun accountMap(a: com.dsrv.wallet.sdk.AccountInfo) = mapOf(
        "accountId" to a.accountId,
        "label" to a.label,
        "addresses" to a.addresses.map {
            mapOf(
                "accountId" to it.accountId,
                "addressId" to it.addressId,
                "address" to it.address,
                "publicKey" to it.publicKey,
                "label" to it.label,
                "chainType" to it.chainType
            )
        }
    )

    private fun chainMap(c: com.dsrv.wallet.sdk.ChainInfo) = mapOf(
        "chainId" to c.chainId, "name" to c.name,
        "chainType" to c.chainType, "networkType" to c.networkType
    )

    private fun restoredMap(r: com.dsrv.wallet.sdk.RestoredKey) =
        mapOf("address" to r.address, "success" to r.success, "error" to r.error)

    private fun chainTxResultMap(r: com.dsrv.wallet.sdk.ChainTxResult) = mapOf(
        "chainId" to r.chainId,
        "outcome" to r.outcome,
        "txHash" to r.txHash,
        "errorMessage" to r.errorMessage,
    )

    private fun challengeReqMap(r: ChallengeRequest) = mapOf(
        "sdkId" to r.sdkId,
        "appId" to r.appId,
        "credentialType" to r.userCredential.type.name.let { enumToWire(it) },
        "credentialValue" to r.userCredential.value,
        "provider" to r.userCredential.provider,
        "deviceInfo" to mapOf(
            "publicKey" to r.deviceInfo.publicKey,
            "model" to r.deviceInfo.model,
            "osVersion" to r.deviceInfo.osVersion,
            "isVirtual" to r.deviceInfo.isVirtual,
            "platform" to "ANDROID"
        )
    )

    private fun wireToEnum(wire: String) = when (wire) {
        "USER_ID" -> "USER_ID"
        "OAUTH_TOKEN" -> "OAUTH_TOKEN"
        "IDP_TOKEN" -> "IDP_TOKEN"
        else -> "USER_ID"
    }

    private fun enumToWire(name: String) = name // CredentialType.name 이 곧 wire 값

    private inline fun <reified T> MethodCall.arg(key: String): T =
        argument<T>(key) ?: throw IllegalArgumentException("missing arg: $key")
}
